documentation/intrinsics.md

```diff
@@ -0,0 +1,658 @@
+the actual arguments are arrays, the arguments must be conformable, and
+the result is also an array.
+Scalar arguments are expanded when the arguments are not all scalars.
+
+## Elemental intrinsic functions that may have unrestricted specific procedures
+
+When an elemental intrinsic function is documented here as having an
+_unrestricted specific name_, that name may be passed as an actual
+argument, used as the target of a procedure pointer, appear in
+a generic interface, and be otherwise used as if it were an external
+procedure.
+An `INTRINSIC` statement or attribute may have to be applied to an
+unrestricted specific name to enable such usage.
+
+In such usage, the instance of the function that accepts and returns
+values of the default kinds of the intrinsic types is used.
+A Fortran `INTERFACE` could be written to define each of
+these unrestricted specific intrinsic function names.
+
+Calls to dummy arguments and procedure pointers that correspond to these
+specific names must pass only scalar actual argument values.
+
+No other intrinsic function name can be passed as an actual argument,
+used as a pointer target, or appear in a generic interface.
+
+### Trigonometric elemental intrinsic functions, generic and (mostly) specific
+All of these functions can be used as unrestricted specific names.
+
+```
+ACOS(REAL(k) X) -> REAL(k)
+ASIN(REAL(k) X) -> REAL(k)
+ATAN(REAL(k) X) -> REAL(k)
+ATAN(REAL(k) Y, REAL(k) X) -> REAL(k) = ATAN2(Y, X)
+ATAN2(REAL(k) Y, REAL(k) X) -> REAL(k)
+COS(REAL(k) X) -> REAL(k)
+COSH(REAL(k) X) -> REAL(k)
+SIN(REAL(k) X) -> REAL(k)
+SINH(REAL(k) X) -> REAL(k)
+TAN(REAL(k) X) -> REAL(k)
+TANH(REAL(k) X) -> REAL(k)
+```
+
+These `COMPLEX` versions of some of those functions, and the
+inverse hyperbolic functions, cannot be used as specific names.
+```
+ACOS(COMPLEX(k) X) -> COMPLEX(k)
+ASIN(COMPLEX(k) X) -> COMPLEX(k)
+ATAN(COMPLEX(k) X) -> COMPLEX(k)
+ACOSH(REAL(k) X) -> REAL(k)
+ACOSH(COMPLEX(k) X) -> COMPLEX(k)
+ASINH(REAL(k) X) -> REAL(k)
+ASINH(COMPLEX(k) X) -> COMPLEX(k)
+ATANH(REAL(k) X) -> REAL(k)
+ATANH(COMPLEX(k) X) -> COMPLEX(k)
+COS(COMPLEX(k) X) -> COMPLEX(k)
+COSH(COMPLEX(k) X) -> COMPLEX(k)
+SIN(COMPLEX(k) X) -> COMPLEX(k)
+SINH(COMPLEX(k) X) -> COMPLEX(k)
+TAN(COMPLEX(k) X) -> COMPLEX(k)
+TANH(COMPLEX(k) X) -> COMPLEX(k)
+```
+
+### Non-trigonometric elemental intrinsic functions, generic and specific
+These functions *can* be used as unrestricted specific names.
+```
+ABS(REAL(k) A) -> REAL(k) = SIGN(A, 0.0)
+AIMAG(COMPLEX(k) Z) -> REAL(k) = Z%IM
+AINT(REAL(k) A [, KIND=k ]) -> REAL(KIND)
+ANINT(REAL(k) A [, KIND=k ] -> REAL(KIND)
+CONJG(COMPLEX(k) Z) -> COMPLEX(k) = CMPLX(Z%RE, -Z%IM)
+DIM(REAL(k) X, REAL(k) Y) -> REAL(k) = X-MIN(X,Y)
+DPROD(default REAL X, default REAL Y) -> DOUBLE PRECISION = DBLE(X)*DBLE(Y)
+EXP(REAL(k) X) -> REAL(k)
+INDEX(CHARACTER(k) STRING, CHARACTER(k) SUBSTRING [, LOGICAL(any) BACK, KIND=KIND(0) ]) -> INTEGER(KIND)
```

Comment:
I assume that any argument with a default value is optional. So even if you specify `BACK` here you don't have to specify `KIND`? Can we write this as:
`INDEX(CHARACTER(k) STRING, CHARACTER(k) SUBSTRING, LOGICAL(any) BACK=.false., KIND=KIND(0))`
and is that any clearer?

It seems like we could eliminate the square bracket syntax for arguments with defaults.

Reply:
Done where possible.  Since argument order matters when calls do not have keywords, some arguments with default values remain in brackets.